File: v4/rust/iprange-livedb/src/publication/replacement_resolver.rs
//! Restart resolution for one exact replacement publication.

use crate::cancellation::CancellationToken;
use crate::error::ErrorCode;

use super::result_builder::{coordination_access, first_problem, record_cancellation, LaterResult};
use super::{
    arm, conflict, require_no_later, reservation_identity, ArmFailure, BaseResolution, Mode,
};
use crate::publication::attempt;
use crate::publication::cleanup::{self, OutputOwner};
use crate::publication::namespace::Destination;
use crate::publication::output::{PreparedOutput, ResumedOutput};
use crate::publication::problem::Problem;
use crate::publication::replacement::PreviousMain;
use crate::publication::replacement_inspection::{self, Content, Inspected, Location, Pair};
use crate::publication::reservation::{Header, Policy, State};
use crate::publication::reservation_inspection;
use crate::publication::result::{
    AccessPolicy, ArtifactKind, DestinationContent, FinalState, NameSlot, PublicationResult,
    PublicationStatus,
};

pub(super) fn dispatch(
    base: BaseResolution,
    mode: Mode,
    cancellation: &CancellationToken,
) -> Result<PublicationResult, Problem> {
    unlock(&base.exact)?;
    unlock(&base.later)?;
    let pair = replacement_inspection::inspect(&base.destination, base.header, cancellation)?;
    relock(&base.exact, &base.destination, cancellation)?;
    relock(&base.later, &base.destination, cancellation)?;
    match pair.main.as_ref().map(|main| main.content) {
        Some(Content::Previous) => {
            require_no_later(&base.later)?;
            resolve_previous(base, pair, mode, cancellation)
        }
        Some(Content::Desired) => resolve_desired(base, pair, mode, cancellation),
        Some(Content::Other) => {
            require_no_later(&base.later)?;
            resolve_other(base, pair, cancellation)
        }
        None => Err(conflict(
            "replacement cannot legitimately leave the destination absent",
        )),
    }
}

fn resolve_previous(
    base: BaseResolution,
    pair: Pair,
    mode: Mode,
    cancellation: &CancellationToken,
) -> Result<PublicationResult, Problem> {
    match mode {
        Mode::Complete => complete_previous(base, pair, cancellation),
        Mode::Remove => remove_previous(base, pair, cancellation),
    }
}

fn complete_previous(
    mut base: BaseResolution,
    mut pair: Pair,
    cancellation: &CancellationToken,
) -> Result<PublicationResult, Problem> {
    let reservation = base
        .exact
        .take()
        .ok_or_else(|| unresolvable("replacement completion requires its exact reservation"))?;
    let main = pair
        .main
        .take()
        .ok_or_else(|| conflict("replacement previous destination disappeared"))?;
    let private = pair
        .private
        .take()
        .ok_or_else(|| unresolvable("replacement completion requires its prepared output"))?;
    require_previous(&main, base.header)?;
    require_output(&private, base.header)?;
    let previous = PreviousMain {
        file: main.file,
        mapping: main
            .mapping
            .expect("finished replacement inspection retains its mapping"),
        identity: main.identity,
        byte_length: main.byte_length,
        sha512: main.sha512,
    };
    let output = PreparedOutput::resume_replacement(
        base.destination,
        base.header.attempt_id,
        ResumedOutput {
            file: private.file,
            mapping: private
                .mapping
                .expect("desired replacement output retains its mapping"),
            identity: private.identity,
            meta: private
                .meta
                .expect("desired replacement output has selected metadata"),
            byte_length: private.byte_length,
            sha512: private.sha512,
        },
        previous,
        base.header.policy,
    )
    .map_err(|error| Problem::output(&error))?;
    cancellation.check().map_err(|error| Problem::sdk(&error))?;
    let reservation_identity = reservation.identity;
    let result = match arm(reservation, &output) {
        Ok(armed) => attempt::resume_armed(base.seed, output, armed),
        Err(ArmFailure::Problem(problem)) => return Err(problem),
        Err(ArmFailure::Unknown(problem)) => {
            attempt::outcome_unknown(base.seed, reservation_identity, problem)
        }
    };
    Ok(record_cancellation(result, cancellation))
}

fn remove_previous(
    base: BaseResolution,
    pair: Pair,
    cancellation: &CancellationToken,
) -> Result<PublicationResult, Problem> {
    let main = pair.main.as_ref().expect("dispatch selected previous main");
    require_previous(main, base.header)?;
    resolve_not_desired(base, pair, DestinationContent::Previous, cancellation)
}

fn resolve_desired(
    mut base: BaseResolution,
    pair: Pair,
    mode: Mode,
    cancellation: &CancellationToken,
) -> Result<PublicationResult, Problem> {
    if base.header.policy == Policy::ReplaceExistingNoRollback && mode == Mode::Remove {
        return Err(unresolvable(
            "no-rollback replacement cannot restore a discarded destination",
        ));
    }
    let main = pair.main.as_ref().expect("dispatch selected desired main");
    synchronize(main, &base.destination, cancellation)?;
    let (output, foreign_private) = desired_cleanup(pair.private.as_ref(), base.header);
    let reservation = base.exact.as_ref().map(super::reservation_owner);
    let mut summary =
        cleanup::discard_recovered(&mut base.seed, &base.destination, output, reservation);
    if let Some(private) = foreign_private {
        summary.artifacts.push(base.seed.artifact(
            ArtifactKind::PrivateOutput,
            NameSlot::PrivateOutput,
            Some(private.identity),
            conflict("private replacement artifact does not match recorded ownership"),
        ));
    }
    let verified = main
        .verify(&base.destination, &CancellationToken::new())
        .and_then(|()| super::verify_later(base.later.as_ref(), &base.destination));
    let cause = verified
        .as_ref()
        .err()
        .cloned()
        .or_else(|| first_problem(&summary.artifacts));
    let publication = if verified.is_ok() {
        PublicationStatus::Published
    } else {
        PublicationStatus::OutcomeUnknown
    };
    let result = base
        .seed
        .result_with_housekeeping(
            FinalState {
                reservation_identity: reservation_identity(base.header),
                main_namespace_may_have_been_attempted: attempted(base.header.state),
                publication,
                destination_content: if publication == PublicationStatus::Published {
                    DestinationContent::Desired
                } else {
                    DestinationContent::Unclassified
                },
                main_access_policy: if verified.is_ok() {
                    main.access
                } else {
                    AccessPolicy::Unclassified
                },
                coordination_access_policy: if verified.is_ok() {
                    coordination_access(&summary, base.exact.as_ref(), base.later.as_ref())
                } else {
                    AccessPolicy::Unclassified
                },
            },
            summary.artifacts,
            summary.housekeeping,
            summary.visible_housekeeping,
            cause,
        )
        .with_later(base.later.as_ref());
    Ok(record_cancellation(result, cancellation))
}

fn resolve_other(
    base: BaseResolution,
    pair: Pair,
    cancellation: &CancellationToken,
) -> Result<PublicationResult, Problem> {
    resolve_not_desired(base, pair, DestinationContent::Other, cancellation)
}

fn resolve_not_desired(
    mut base: BaseResolution,
    pair: Pair,
    content: DestinationContent,
    cancellation: &CancellationToken,
) -> Result<PublicationResult, Problem> {
    let main = pair
        .main
        .as_ref()
        .expect("dispatch selected a non-desired main");
    synchronize(main, &base.destination, cancellation)?;
    let output = removable_output(pair.private.as_ref(), base.header)?;
    let reservation = base.exact.as_ref().map(super::reservation_owner);
    let summary =
        cleanup::discard_recovered(&mut base.seed, &base.destination, output, reservation);
    let verified = main.verify(&base.destination, &CancellationToken::new());
    let cause = verified
        .as_ref()
        .err()
        .cloned()
        .or_else(|| first_problem(&summary.artifacts));
    let publication = if verified.is_ok() {
        PublicationStatus::NotPublished
    } else {
        PublicationStatus::OutcomeUnknown
    };
    let result = base.seed.result_with_housekeeping(
        FinalState {
            reservation_identity: reservation_identity(base.header),
            main_namespace_may_have_been_attempted: attempted(base.header.state),
            publication,
            destination_content: if verified.is_ok() {
                content
            } else {
                DestinationContent::Unclassified
            },
            main_access_policy: if verified.is_ok() {
                main.access
            } else {
                AccessPolicy::Unclassified
            },
            coordination_access_policy: if verified.is_ok() {
                coordination_access(&summary, base.exact.as_ref(), None)
            } else {
                AccessPolicy::Unclassified
            },
        },
        summary.artifacts,
        summary.housekeeping,
        summary.visible_housekeeping,
        cause,
    );
    Ok(record_cancellation(result, cancellation))
}

fn removable_output(
    private: Option<&Inspected>,
    header: Header,
) -> Result<Option<OutputOwner<'_>>, Problem> {
    match private {
        None => Ok(None),
        Some(private)
            if private.content == Content::Desired
                && private.identity.encode() == header.output_identity =>
        {
            Ok(Some(owner(private)))
        }
        Some(_) => Err(conflict(
            "private replacement artifact does not match the prepared output",
        )),
    }
}

fn desired_cleanup(
    private: Option<&Inspected>,
    header: Header,
) -> (Option<OutputOwner<'_>>, Option<&Inspected>) {
    match private {
        None => (None, None),
        Some(private)
            if (private.content == Content::Desired
                && private.identity.encode() == header.output_identity)
                || (private.content == Content::Previous
                    && header.previous.is_some_and(|previous| {
                        private.identity.encode() == previous.identity
                    })) =>
        {
            (Some(owner(private)), None)
        }
        Some(private) => (None, Some(private)),
    }
}

fn owner(entry: &Inspected) -> OutputOwner<'_> {
    OutputOwner {
        file: &entry.file,
        identity: entry.identity,
        name: &entry.name,
    }
}

fn require_output(output: &Inspected, header: Header) -> Result<(), Problem> {
    if output.location != Location::Private
        || output.content != Content::Desired
        || output.identity.encode() != header.output_identity
        || output.access != AccessPolicy::CreatorOnly
    {
        return Err(unresolvable(
            "replacement prepared output does not match its reservation",
        ));
    }
    Ok(())
}

fn require_previous(previous: &Inspected, header: Header) -> Result<(), Problem> {
    let expected = header
        .previous
        .expect("replacement header contains previous evidence");
    if previous.content != Content::Previous || previous.identity.encode() != expected.identity {
        return Err(conflict(
            "replacement destination no longer matches previous evidence",
        ));
    }
    Ok(())
}

fn synchronize(
    main: &Inspected,
    destination: &Destination,
    cancellation: &CancellationToken,
) -> Result<(), Problem> {
    cancellation.check().map_err(|error| Problem::sdk(&error))?;
    main.file
        .sync_all()
        .map_err(crate::error::Error::from)
        .map_err(|error| Problem::sdk(&error))?;
    main.verify(destination, cancellation)
}

fn unlock(reservation: &Option<reservation_inspection::Inspected>) -> Result<(), Problem> {
    if let Some(reservation) = reservation {
        reservation.unlock_operation()?;
    }
    Ok(())
}

fn relock(
    reservation: &Option<reservation_inspection::Inspected>,
    destination: &Destination,
    cancellation: &CancellationToken,
) -> Result<(), Problem> {
    if let Some(reservation) = reservation {
        reservation.relock_operation(destination, cancellation)?;
    }
    Ok(())
}

const fn attempted(state: State) -> bool {
    matches!(state, State::MainMayHaveBeenAttempted)
}

const fn unresolvable(detail: &'static str) -> Problem {
    Problem::new(ErrorCode::Unresolvable, None, detail)
}

File: v4/go/internal/publication/replacement_resolver.go
//go:build !windows

// Restart resolution for one exact replacement publication (Rust
// publication/replacement_resolver.rs): the exact reservation and
// the pair inspection unlock first (so the two-inode lifetime locks
// can take their role order), the main and private outputs classify
// against the recorded previous and desired evidence, and the
// previous is completed to desired, restored, or refused with the
// exact cleanup ledger. The inspected pair and the base reservations
// close exactly where Rust drops them.

package publication

// replacementDispatch routes one replacement resolution (Rust
// replacement_resolver::dispatch): the operation lock is released
// for the pair inspection and re-acquired, the main class selects
// the arm, and the destination directory closes at the scope end
// (Rust drops the BaseResolution destination).
func replacementDispatch(base baseResolution, mode resolveMode, check func() error) (result PublicationResult, err error) {
	defer base.destination.directory().Close()
	if err := unlockReplacementReservation(base.exact); err != nil {
		closeInspectedReservation(base.exact)
		closeInspectedReservation(base.later)
		return PublicationResult{}, resolverProblem(err)
	}
	if err := unlockReplacementReservation(base.later); err != nil {
		closeInspectedReservation(base.exact)
		closeInspectedReservation(base.later)
		return PublicationResult{}, resolverProblem(err)
	}
	pair, err := inspectReplacementPair(base.destination, base.header, check)
	if err != nil {
		closeInspectedReservation(base.exact)
		closeInspectedReservation(base.later)
		return PublicationResult{}, err
	}
	if err := relockReplacementReservation(base.exact, base.destination, check); err != nil {
		pair.main.closeIfNonNil()
		pair.private.closeIfNonNil()
		closeInspectedReservation(base.exact)
		closeInspectedReservation(base.later)
		return PublicationResult{}, resolverProblem(err)
	}
	if err := relockReplacementReservation(base.later, base.destination, check); err != nil {
		pair.main.closeIfNonNil()
		pair.private.closeIfNonNil()
		closeInspectedReservation(base.exact)
		closeInspectedReservation(base.later)
		return PublicationResult{}, resolverProblem(err)
	}
	switch {
	case pair.main != nil && pair.main.content == replacementContentPrevious:
		if err := requireNoLater(base.later); err != nil {
			pair.main.closeIfNonNil()
			pair.private.closeIfNonNil()
			closeInspectedReservation(base.exact)
			closeInspectedReservation(base.later)
			return PublicationResult{}, err
		}
		return resolvePreviousReplacement(base, pair, mode, check)
	case pair.main != nil && pair.main.content == replacementContentDesired:
		return resolveDesiredReplacement(base, pair, mode, check)
	case pair.main != nil:
		if err := requireNoLater(base.later); err != nil {
			pair.main.closeIfNonNil()
			pair.private.closeIfNonNil()
			closeInspectedReservation(base.exact)
			closeInspectedReservation(base.later)
			return PublicationResult{}, err
		}
		return resolveOtherReplacement(base, pair, check)
	default:
		pair.main.closeIfNonNil()
		pair.private.closeIfNonNil()
		closeInspectedReservation(base.exact)
		closeInspectedReservation(base.later)
		return PublicationResult{}, conflictProblem("replacement cannot legitimately leave the destination absent")
	}
}

// resolvePreviousReplacement splits by mode (Rust resolve_previous).
func resolvePreviousReplacement(base baseResolution, pair replacementPair, mode resolveMode, check func() error) (PublicationResult, error) {
	if mode == resolveModeComplete {
		return completePreviousReplacement(base, pair, check)
	}
	return removePreviousReplacement(base, pair, check)
}

// completePreviousReplacement resumes the interrupted replacement
// from its exact previous main, prepared output, and reservation
// (Rust complete_previous): the pair entries move into the previous
// main and the prepared output, the reservation arms, and the attempt
// machine retires the displaced previous.
func completePreviousReplacement(base baseResolution, pair replacementPair, check func() error) (PublicationResult, error) {
	if base.exact == nil {
		pair.main.closeIfNonNil()
		pair.private.closeIfNonNil()
		return PublicationResult{}, unresolvable("replacement completion requires its exact reservation")
	}
	reservation := base.exact
	base.exact = nil
	if pair.main == nil {
		pair.private.closeIfNonNil()
		closeInspectedReservation(reservation)
		return PublicationResult{}, conflictProblem("replacement previous destination disappeared")
	}
	main := pair.main
	pair.main = nil
	if pair.private == nil {
		_ = main.Close()
		closeInspectedReservation(reservation)
		return PublicationResult{}, unresolvable("replacement completion requires its prepared output")
	}
	private := pair.private
	pair.private = nil
	if err := requirePreviousReplacement(main, base.header); err != nil {
		_ = main.Close()
		_ = private.Close()
		closeInspectedReservation(reservation)
		return PublicationResult{}, err
	}
	if err := requireOutputReplacement(private, base.header); err != nil {
		_ = main.Close()
		_ = private.Close()
		closeInspectedReservation(reservation)
		return PublicationResult{}, err
	}
	previous := &previousMain{
		file:       main.file,
		mapping:    main.mapping,
		identity:   main.identity,
		byteLength: main.byteLength,
		sha512:     main.sha512,
	}
	main.file = nil
	main.mapping = nil
	output, err := resumePreparedOutputReplacement(base.destination, base.header, private, previous)
	if err != nil {
		// The constructor error path already closed the inspected
		// private artifact; the previous main is still owned here
		// and must be closed by this arm (Rust drops the moved
		// PreviousMain inside the failed construction).
		_ = previous.Close()
		closeInspectedReservation(reservation)
		return PublicationResult{}, outputProblem(err)
	}
	private.file = nil
	private.mapping = nil
	if err := checkCancellation(check); err != nil {
		_ = output.Close()
		_ = main.Close()
		closeInspectedReservation(reservation)
		return PublicationResult{}, err
	}
	reservationIdentity := reservation.identity
	armed, armFailure := arm(reservation, output)
	reservation = nil
	if armFailure != nil {
		_ = output.Close()
		if armFailure.unknown {
			return recordCancellation(outcomeUnknown(base.seed, reservationIdentity, armFailure.problem), check), nil
		}
		return PublicationResult{}, armFailure.problem
	}
	result := resumeArmed(base.seed, output, armed)
	_ = output.Close()
	return recordCancellation(result, check), nil
}

// removePreviousReplacement removes the interrupted replacement and
// restores the previous main (Rust remove_previous).
func removePreviousReplacement(base baseResolution, pair replacementPair, check func() error) (PublicationResult, error) {
	if pair.main == nil {
		pair.private.closeIfNonNil()
		closeInspectedReservation(base.exact)
		closeInspectedReservation(base.later)
		return PublicationResult{}, conflictProblem("replacement previous destination disappeared")
	}
	if err := requirePreviousReplacement(pair.main, base.header); err != nil {
		pair.main.closeIfNonNil()
		pair.private.closeIfNonNil()
		closeInspectedReservation(base.exact)
		closeInspectedReservation(base.later)
		return PublicationResult{}, err
	}
	return resolveNotDesiredReplacement(base, pair, DestinationContentPrevious, check)
}

// resolveDesiredReplacement resolves a replacement whose main is the
// desired bytes (Rust resolve_desired): the no-rollback remove mode
// is refused, the private artifact is either the recorded output
// (discarded) or a foreign residue (artifact, never removed), and the
// result carries the later-canonical observation.
func resolveDesiredReplacement(base baseResolution, pair replacementPair, mode resolveMode, check func() error) (PublicationResult, error) {
	if base.header.policy == reservationPolicyReplaceExistingNoRollback && mode == resolveModeRemove {
		pair.main.closeIfNonNil()
		pair.private.closeIfNonNil()
		closeInspectedReservation(base.exact)
		closeInspectedReservation(base.later)
		return PublicationResult{}, unresolvable("no-rollback replacement cannot restore a discarded destination")
	}
	if pair.main == nil {
		pair.private.closeIfNonNil()
		closeInspectedReservation(base.exact)
		closeInspectedReservation(base.later)
		return PublicationResult{}, conflictProblem("replacement desired destination disappeared")
	}
	if err := synchronizeReplacement(pair.main, base.destination, check); err != nil {
		pair.main.closeIfNonNil()
		pair.private.closeIfNonNil()
		closeInspectedReservation(base.exact)
		closeInspectedReservation(base.later)
		// Rust propagates every synchronize failure with ? (no
		// outcome-unknown conversion on the replacement arms).
		return PublicationResult{}, err
	}
	owner, removable, foreign := desiredCleanupReplacement(pair.private, base.header)
	var output *outputOwner
	if removable {
		output = &owner
	}
	s := &base.seed
	summary := discardRecovered(s, base.destination, output, reservationOwnerOf(base.exact))
	if foreign != nil {
		summary.artifacts.push(s.artifact(ArtifactPrivateOutput, nameSlotPrivateOutput, identityOptional{present: true, identity: foreign.identity}, conflictProblem("private replacement artifact does not match recorded ownership")))
	}
	verified := resolverProblem(pair.main.verify(base.destination, noopCheck))
	if verified == nil {
		verified = resolverProblem(verifyLater(base.later, base.destination))
	}
	cause := verified
	if cause == nil {
		cause = firstProblem(&summary.artifacts)
	}
	publication := PublicationOutcomeUnknown
	if verified == nil {
		publication = PublicationPublished
	}
	state := finalState{
		reservationIdentity:               reservationIdentityOf(base.header),
		mainNamespaceMayHaveBeenAttempted: attemptedReplacement(base.header.state),
		publication:                       publication,
		destinationContent:                DestinationContentUnclassified,
		mainAccessPolicy:                  AccessPolicyUnclassified,
		coordinationAccessPolicy:          AccessPolicyUnclassified,
	}
	if verified == nil {
		state.destinationContent = DestinationContentDesired
		state.mainAccessPolicy = pair.main.access
		state.coordinationAccessPolicy = coordinationAccess(summary, base.exact, base.later)
	}
	result := s.resultWithHousekeeping(state, summary.artifacts, summary.housekeeping, summary.visibleHousekeeping, cause)
	result = withLater(result, base.later)
	pair.main.closeIfNonNil()
	pair.private.closeIfNonNil()
	closeInspectedReservation(base.exact)
	closeInspectedReservation(base.later)
	return recordCancellation(result, check), nil
}

// resolveOtherReplacement resolves a foreign main (Rust
// resolve_other -> resolve_not_desired).
func resolveOtherReplacement(base baseResolution, pair replacementPair, check func() error) (PublicationResult, error) {
	return resolveNotDesiredReplacement(base, pair, DestinationContentOther, check)
}

// resolveNotDesiredReplacement removes the interrupted replacement
// artifacts and restores the classified main (Rust
// resolve_not_desired: the private output is removed only when it is
// the recorded prepared output).
func resolveNotDesiredReplacement(base baseResolution, pair replacementPair, content DestinationContent, check func() error) (PublicationResult, error) {
	if pair.main == nil {
		pair.private.closeIfNonNil()
		closeInspectedReservation(base.exact)
		closeInspectedReservation(base.later)
		return PublicationResult{}, conflictProblem("replacement non-desired destination disappeared")
	}
	if err := synchronizeReplacement(pair.main, base.destination, check); err != nil {
		pair.main.closeIfNonNil()
		pair.private.closeIfNonNil()
		closeInspectedReservation(base.exact)
		closeInspectedReservation(base.later)
		// Rust propagates every synchronize failure with ? (no
		// outcome-unknown conversion on the replacement arms).
		return PublicationResult{}, err
	}
	owner, removable, err := removableReplacementOutput(pair.private, base.header)
	if err != nil {
		pair.main.closeIfNonNil()
		pair.private.closeIfNonNil()
		closeInspectedReservation(base.exact)
		closeInspectedReservation(base.later)
		return PublicationResult{}, err
	}
	var output *outputOwner
	if removable {
		output = &owner
	}
	s := &base.seed
	summary := discardRecovered(s, base.destination, output, reservationOwnerOf(base.exact))
	verified := resolverProblem(pair.main.verify(base.destination, noopCheck))
	cause := verified
	if cause == nil {
		cause = firstProblem(&summary.artifacts)
	}
	publication := PublicationOutcomeUnknown
	if verified == nil {
		publication = PublicationNotPublished
	}
	state := finalState{
		reservationIdentity:               reservationIdentityOf(base.header),
		mainNamespaceMayHaveBeenAttempted: attemptedReplacement(base.header.state),
		publication:                       publication,
		destinationContent:                DestinationContentUnclassified,
		mainAccessPolicy:                  AccessPolicyUnclassified,
		coordinationAccessPolicy:          AccessPolicyUnclassified,
	}
	if verified == nil {
		state.destinationContent = content
		state.mainAccessPolicy = pair.main.access
		state.coordinationAccessPolicy = coordinationAccess(summary, base.exact, nil)
	}
	result := s.resultWithHousekeeping(state, summary.artifacts, summary.housekeeping, summary.visibleHousekeeping, cause)
	pair.main.closeIfNonNil()
	pair.private.closeIfNonNil()
	closeInspectedReservation(base.exact)
	closeInspectedReservation(base.later)
	return recordCancellation(result, check), nil
}

// removableReplacementOutput keeps the prepared-output owner only
// when the private artifact is exactly the recorded output (Rust
// removable_output: any other private artifact is the conflict
// class). The owner returns by value; the caller takes its address
// only within its own frame.
func removableReplacementOutput(private *inspectedReplacement, header reservationHeader) (outputOwner, bool, error) {
	if private == nil {
		return outputOwner{}, false, nil
	}
	if private.content == replacementContentDesired && reservationIdentityBytes(private.identity) == header.outputIdentity {
		return replacementOwnerOf(private), true, nil
	}
	return outputOwner{}, false, conflictProblem("private replacement artifact does not match the prepared output")
}

// desiredCleanupReplacement separates the recorded private output
// from a foreign private residue (Rust desired_cleanup: the recorded
// output or the recorded previous inode is removed; anything else is
// left as residue with its artifact). The owner returns by value; the
// caller takes its address only within its own frame.
func desiredCleanupReplacement(private *inspectedReplacement, header reservationHeader) (outputOwner, bool, *inspectedReplacement) {
	if private == nil {
		return outputOwner{}, false, nil
	}
	if private.content == replacementContentDesired && reservationIdentityBytes(private.identity) == header.outputIdentity {
		return replacementOwnerOf(private), true, nil
	}
	if private.content == replacementContentPrevious && header.previousPresent && reservationIdentityBytes(private.identity) == header.previous.identity {
		return replacementOwnerOf(private), true, nil
	}
	return outputOwner{}, false, private
}

// replacementOwnerOf builds the cleanup owner of one inspected
// replacement (Rust owner).
func replacementOwnerOf(entry *inspectedReplacement) outputOwner {
	return outputOwner{
		file:     entry.file,
		identity: entry.identity,
		name:     entry.name,
	}
}

// requireOutputReplacement refuses one prepared output that no
// longer matches the reservation (Rust require_output: private
// placement, desired content, recorded identity, creator-only
// access).
func requireOutputReplacement(output *inspectedReplacement, header reservationHeader) error {
	if output.location != outputLocationPrivate ||
		output.content != replacementContentDesired ||
		reservationIdentityBytes(output.identity) != header.outputIdentity ||
		output.access != AccessPolicyCreatorOnly {
		return unresolvable("replacement prepared output does not match its reservation")
	}
	return nil
}

// requirePreviousReplacement refuses one previous main that no
// longer matches the recorded evidence (Rust require_previous).
func requirePreviousReplacement(previous *inspectedReplacement, header reservationHeader) error {
	if !header.previousPresent {
		return conflictProblem("replacement previous evidence is missing")
	}
	if previous.content != replacementContentPrevious || reservationIdentityBytes(previous.identity) != header.previous.identity {
		return conflictProblem("replacement destination no longer matches previous evidence")
	}
	return nil
}

// synchronizeReplacement durably flushes one inspected main and
// re-proves it (Rust replacement_resolver::synchronize).
func synchronizeReplacement(main *inspectedReplacement, destination *destination, check func() error) error {
	if err := checkCancellation(check); err != nil {
		return err
	}
	if err := syncOutputFile(main.file); err != nil {
		return resolverProblem(err)
	}
	return main.verify(destination, check)
}

// unlockReplacementReservation releases the operation lock of one
// base reservation before the pair inspection (Rust unlock).
func unlockReplacementReservation(reservation *inspectedReservation) error {
	if reservation == nil {
		return nil
	}
	return reservation.unlockOperation()
}

// relockReplacementReservation re-acquires the operation lock and
// re-proves one base reservation after the pair inspection (Rust
// relock).
func relockReplacementReservation(reservation *inspectedReservation, destination *destination, check func() error) error {
	if reservation == nil {
		return nil
	}
	return reservation.relockOperation(destination, check)
}

// attemptedReplacement reports whether one header state marks the
// main namespace attempted (Rust attempted).
func attemptedReplacement(state reservationState) bool {
	return state == reservationStateMainMayHaveBeenAttempted
}
